<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Luis Jorge Ahumada Cerda, Dario Castro Gonzalez

**Fecha:** 09 de febrero 2023.

**Expediente** : 728125, 719910.
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: https://github.com/kodharius/ProyectoConjuntoCastroAhumada

# Tarea 2: Clase 4

## Enunciado de tarea (Transcibirlo)
# Tarea 2:

1. Imagine una rutina contable utilizada en una librería. Funciona en una lista con sublistas, que se ven así:
![imagen.png](attachment:imagen.png)
Escriba un programa en Python, que devuelve una lista con tuplas de tamaño 2. Cada tupla consiste en el número de pedido y el costo total del pedido. El producto debe aumentarse en 10€ si el valor de la orden es inferior a 100,00 €. 
Escribe un programa Python usando unicamente las funciones **lambda y map**.

`orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]`
           
2. La misma librería, pero esta vez trabajamos en una lista diferente. Las sublistas de nuestras listas se ven así: 
[número de orden, (número de artículo, cantidad, precio por unidad), ... (número de artículo, cantidad, precio por unidad)] 

`orders = [[1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
	       [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
	       [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
           [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]`

Escriba un programa que devuelva una lista de dos tuplas que tengan la información de (número de pedido, cantidad total de pedido). Utilice la función `Reduce`. Recuerde el precio del pedido del producto debe aumentarse en 10€ si el valor de la orden es inferior a 100,00 €.

3. Resolver los anteriores ejercicios usando pandas (DataFrame)

## Solucion 1

In [1]:
########### Solución 1
orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]

# El resultado debe ser como muestro en el resultado de esta celda


In [2]:
import pandas as pd

orders = [["34587", "Learning Python, Mark Lutz", 4, 40.95], 
          ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
          ["77226", "Head First Python, Paul Barry", 3, 32.95], 
          ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]
# solucion  estudiante 1 
df = pd.DataFrame(orders, columns=['order_number', 'item', 'quantity', 'price'])
df['total_price'] = df['quantity'] * df['price']
df['total_price'] = df['total_price'].apply(lambda x: x + 10 if x < 100 else x)
df = df[['order_number', 'total_price']]
result = df.values.tolist()
print(result)

[['34587', 163.8], ['98762', 284.0], ['77226', 108.85000000000001], ['88112', 84.97]]


In [3]:
#solucion  estudiante 2 
total_cost = lambda x: x[2] * x[3] + 10 if x[2] * x[3] < 100 else x[2] * x[3]
result = list(map(lambda x: (x[0], total_cost(x)), orders))
print(result)

[('34587', 163.8), ('98762', 284.0), ('77226', 108.85000000000001), ('88112', 84.97)]


### Respuesta planteada en el ejercicio usando (markdown)
El costo de ordenes seria el segundo numero de la tupla y el primer numero es el numero de orden [('34587', 163.8), ('98762', 284.0), ('77226', 108.85000000000001), ('88112', 84.97)]

## Solucion 2

In [4]:
import pandas as np
from functools import reduce

orders = [[1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
           [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
           [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
           [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]
# estudiante 1 
resultado = []
for order in orders:
    total = reduce(lambda x, y: x + y[1] * y[2], order[1:], 0)
    if total < 100:
        total += 10
    resultado.append([order[0], total])

print(resultado)

[[1, 678.3299999999999], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57]]


In [5]:
# estudiante 2 
resultado = []
for order in orders:
    total = sum([item[1] * item[2] for item in order[1:]])
    total = total + 10 if total < 100 else total
    resultado.append([order[0], total])

print(resultado)


[[1, 678.3299999999999], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57]]


### Respuesta planteada en el ejercicio usando (markdown)
El costo de ordenes seria el segundo numero de la tupla [[1, 678.3299999999999], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57]]

# solucion 3

In [6]:
import pandas as pd

orders = [["34587", "Learning Python, Mark Lutz", 4, 40.95], 
          ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
          ["77226", "Head First Python, Paul Barry", 3, 32.95], 
          ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]
# solucion problema 1 df estudiante 1
df = pd.DataFrame(orders, columns=['order_number', 'item', 'quantity', 'price'])
df['total_price'] = df['quantity'] * df['price']
df['total_price'] = df['total_price'].apply(lambda x: x + 10 if x < 100 else x)
df = df[['order_number', 'total_price']]
result = df.values.tolist()
print(result)

[['34587', 163.8], ['98762', 284.0], ['77226', 108.85000000000001], ['88112', 84.97]]


In [7]:
#solucion estudiante 2 
orders = [["34587", "Learning Python, Mark Lutz", 4, 40.95], 
          ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
          ["77226", "Head First Python, Paul Barry", 3, 32.95], 
          ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]
df = pd.DataFrame(orders, columns=['order_number', 'item', 'quantity', 'price'])
df['total_price'] = df['quantity'] * df['price']
df['total_price'] = df['total_price'].apply(lambda x: x + 10 if x < 100 else x)
df = df[['order_number', 'total_price']]
result = df.values.tolist()
print(result)


[['34587', 163.8], ['98762', 284.0], ['77226', 108.85000000000001], ['88112', 84.97]]


## Solucion 4

In [8]:
#solucion estudiante 1
palabras = 'the notion of a lambda function goes all the way back to the origin of computer science'.split()
res = list(filter(lambda word: len(word) >= 4, palabras))
print(res)

['notion', 'lambda', 'function', 'goes', 'back', 'origin', 'computer', 'science']


In [9]:
#solucion estudiante 2 
palabras = 'the notion of a lambda function goes all the way back to the origin of computer science'.split()
res = [word for word in palabras if len(word) >= 4]
print(res)


['notion', 'lambda', 'function', 'goes', 'back', 'origin', 'computer', 'science']


### Respuesta planteada en el ejercicio usando (markdown)
 La respuesta es ['notion', 'lambda', 'function', 'goes', 'back', 'origin', 'computer', 'science']

## Solucion 5

In [10]:
import numpy as np
np.random.seed(55555)
temperatures = np.random.uniform(25,110,30)
print(temperatures)


[ 35.97387961  99.13294268  92.6178302   99.69549478  63.21119535
  73.90387697  95.50028838  99.21476553  36.98251238 101.45813729
 100.5917418   89.38159435  89.52156589  83.8084752   28.37990868
  68.75395211  45.86132067  36.58307493  56.32208365  33.96219302
  45.29071318  74.86289938  66.51091383  88.19316186  94.26635864
  46.83131624 109.60744472 109.68427993  44.55796499  92.84146259]


In [11]:
#solucion estudiante 1
# Usamos filtros para filtrar las temp deseadas
filtro_temps = list(filter(lambda x: 32 <= x <= 80, temperatures))

# Imprimimos las temp deseadas entre 32 a 80
print("Temperaturas entre 32 a 80: ", filtro_temps)

Temperaturas entre 32 a 80:  [35.97387961379455, 63.211195349983264, 73.90387696517755, 36.982512376595906, 68.75395211338406, 45.86132066984196, 36.58307493022292, 56.322083650154724, 33.9621930206693, 45.29071317816652, 74.86289938282341, 66.51091382521807, 46.831316244170544, 44.55796498523391]


In [12]:
#solucion estudiante 2

import numpy as np
np.random.seed(55555)
temperatures = np.random.uniform(25,110,30)


filtro_temps = [x for x in temperatures if 32 <= x <= 80]

print("Temperaturas entre 32 a 80: ", filtro_temps)


Temperaturas entre 32 a 80:  [35.97387961379455, 63.211195349983264, 73.90387696517755, 36.982512376595906, 68.75395211338406, 45.86132066984196, 36.58307493022292, 56.322083650154724, 33.9621930206693, 45.29071317816652, 74.86289938282341, 66.51091382521807, 46.831316244170544, 44.55796498523391]


### Respuesta planteada en el ejercicio usando (markdown)
 La respuesta es Temperaturas entre 32 a 80:  [35.97387961379455, 63.211195349983264, 73.90387696517755, 36.982512376595906, 68.75395211338406, 45.86132066984196, 36.58307493022292, 56.322083650154724, 33.9621930206693, 45.29071317816652, 74.86289938282341, 66.51091382521807, 46.831316244170544, 44.55796498523391]

## Solucion 6

In [13]:
########### Solución 5
np.random.seed(55555)
point_xy = np.random.randint(0,100,[30,2])
print(point_xy)

[[17 97]
 [22 33]
 [37 46]
 [57 28]
 [11 25]
 [14  2]
 [74 16]
 [12 37]
 [35 21]
 [66 65]
 [86 58]
 [67 81]
 [ 0 48]
 [28 66]
 [76 22]
 [46 47]
 [64  7]
 [88 94]
 [65 13]
 [75 89]
 [23 17]
 [36 63]
 [62 36]
 [64 17]
 [18 19]
 [88 41]
 [21 49]
 [98 36]
 [81 73]
 [62 90]]


In [14]:
from functools import reduce
lista = point_xy.tolist()

# Realiza la suma de los elementos de una lista

# Determinar el máximo de una lista de valores numéricos 
f = reduce(lambda x, y: x if x>y else y, lista)
print('El máximo valor de la lista es:',  f)
f3 = reduce(min, lista)
print('El minimo valor de la lista es:',  f3)


El máximo valor de la lista es: [98, 36]
El minimo valor de la lista es: [0, 48]


In [15]:
import numpy as np

np.random.seed(55555)
point_xy = np.random.randint(0, 100, [30, 2])
print(point_xy)
max_pos = np.argmax(point_xy, axis=1)
max_value = point_xy[max_pos[0], max_pos[1]]
print("El máximo valor de la lista es:", max_value)

min_pos = np.argmin(point_xy, axis=1)
min_value = point_xy[min_pos[0], min_pos[1]]
print("El mínimo valor de la lista es:", min_value)

[[17 97]
 [22 33]
 [37 46]
 [57 28]
 [11 25]
 [14  2]
 [74 16]
 [12 37]
 [35 21]
 [66 65]
 [86 58]
 [67 81]
 [ 0 48]
 [28 66]
 [76 22]
 [46 47]
 [64  7]
 [88 94]
 [65 13]
 [75 89]
 [23 17]
 [36 63]
 [62 36]
 [64 17]
 [18 19]
 [88 41]
 [21 49]
 [98 36]
 [81 73]
 [62 90]]
El máximo valor de la lista es: 33
El mínimo valor de la lista es: 17


### Respuesta planteada en el ejercicio usando (markdown)
 La respuesta es El máximo valor de la lista es: [98, 36]
El minimo valor de la lista es: [0, 48]

In [16]:
#Solucion 2
# Encontrar el valor máximo usando la función reduce
maximo = reduce(lambda x, y: max(x, y), lista)

# Encontrar el valor mínimo usando la función reduce
minimo = reduce(lambda x, y: min(x, y), lista)

# Imprimir los resultados
print(f'El valor máximo es {maximo}.')
print(f'El valor mínimo es {minimo}.')

El valor máximo es [98, 36].
El valor mínimo es [0, 48].
